src/execution/index/art/node.cpp

```diff
@@ -1,59 +1,134 @@
+		break;
+	}
+	case NodeType::N16: {
+		auto n16 = (Node16 *)n;
+		Set(n16->key, 16, n16->children, 16);
+		break;
+	}
+	case NodeType::N48: {
+		auto n48 = (Node48 *)n;
+		Set(n48->child_index, 256, n48->children, 48);
+		break;
+	}
+	case NodeType::N256: {
+		auto n256 = (Node256 *)n;
+		Set(nullptr, 0, n256->children, 256);
+		break;
+	}
+	default:
+		throw InternalException("This is not an Internal ART Node Type");
+	}
+}
+
+void InternalType::Set(uint8_t *key_p, uint16_t key_size_p, SwizzleablePointer *children_p, uint16_t children_size_p) {
+	key = key_p;
+	key_size = key_size_p;
+	children = children_p;
+	children_size = children_size_p;
+}
+
 Node::Node(NodeType type) : count(0), type(type) {
 }
 
 // LCOV_EXCL_START
-Node *Node::GetChild(ART &art, idx_t pos) {
+idx_t Node::GetMin() {
 	D_ASSERT(0);
-	return nullptr;
+	return 0;
 }
 
-void Node::ReplaceChildPointer(idx_t pos, Node *node) {
+Node *Node::GetChild(ART &art, idx_t pos) {
 	D_ASSERT(0);
+	return nullptr;
 }
 
-idx_t Node::GetMin() {
+void Node::ReplaceChildPointer(idx_t pos, Node *node) {
 	D_ASSERT(0);
-	return 0;
 }
 // LCOV_EXCL_STOP
 
-void InternalType::Set(uint8_t *key_p, uint16_t key_size_p, SwizzleablePointer *children_p, uint16_t children_size_p) {
-	key = key_p;
-	key_size = key_size_p;
-	children = children_p;
-	children_size = children_size_p;
+void Node::InsertChild(Node *&node, uint8_t key_byte, Node *new_child) {
+	switch (node->type) {
+	case NodeType::N4:
+		Node4::InsertChild(node, key_byte, new_child);
+		break;
+	case NodeType::N16:
+		Node16::InsertChild(node, key_byte, new_child);
+		break;
+	case NodeType::N48:
+		Node48::InsertChild(node, key_byte, new_child);
+		break;
+	case NodeType::N256:
+		Node256::InsertChild(node, key_byte, new_child);
+		break;
+	default:
+		throw InternalException("Unrecognized leaf type for insert");
+	}
 }
 
-InternalType::InternalType(Node *n) {
-	switch (n->type) {
+void Node::EraseChild(Node *&node, idx_t pos, ART &art) {
+	switch (node->type) {
 	case NodeType::N4: {
-		auto n4 = (Node4 *)n;
-		Set(n4->key, 4, n4->children, 4);
+		Node4::EraseChild(node, pos, art);
 		break;
 	}
 	case NodeType::N16: {
-		auto n16 = (Node16 *)n;
-		Set(n16->key, 16, n16->children, 16);
+		Node16::EraseChild(node, pos, art);
 		break;
 	}
 	case NodeType::N48: {
-		auto n48 = (Node48 *)n;
-		Set(n48->child_index, 256, n48->children, 48);
+		Node48::EraseChild(node, pos, art);
 		break;
 	}
-	case NodeType::N256: {
-		auto n256 = (Node256 *)n;
-		Set(nullptr, 0, n256->children, 256);
+	case NodeType::N256:
+		Node256::EraseChild(node, pos, art);
 		break;
+	default:
+		throw InternalException("Unrecognized leaf type for erase");
+	}
+}
+
+NodeType Node::GetTypeBySize(idx_t size) {
+
+	if (size <= 4) {
+		return NodeType::N4;
+	} else if (size <= 16) {
+		return NodeType::N16;
+	} else if (size <= 48) {
+		return NodeType::N48;
 	}
+	return NodeType::N256;
+}
+
+void Node::New(NodeType &type, Node *&node) {
```

Comment:
Does `type` need to be a reference here?

Reply:
It saves a copy of the type, no? Or how do you mean? Should I make it a `const` reference?

Reply:
Hi, @taniabogatsch nice work! Can you please update this nice blog after merge this PR? THX! 😄https://duckdb.org/2022/07/27/art-storage.html

Reply:
I doubt anything on that blog post needs to change, because nothing changes about how the art is serialized/deserialized as far as I know :)

Reply:
@iwanttobepowerful thanks! I have a few more changes to the ART planned and will for sure keep everyone posted by writing a blog post about them once I'm done. I'll also see if the one on (de)serialization needs updating after! :)

Reply:
nice. Thanks for your attention! :)